### PR TITLE
[WIP] Implementing grpc

### PR DESCRIPTION
**Pull request by ImmanuelSegol**
This is not a complete implementation, I just had some questions. 
I left the questions as comments.

### CODE REVIEW

File: lib/Walli.ts [hunk 1]
```diff
@@ -14,12 +15,13 @@ class Walli {
...
   private btcdClient: BtcdClient;
   private config: Config;
+  private grpcServer: GrpcServer;
 
   constructor(config: ConfigType) {
     this.config = new Config().load(config);
     this.logger = new Logger(this.config.logfile, this.config.loglevel);
-    console.log(this.config);
     this.btcdClient = new BtcdClient(this.config.rpc);
+    this.grpcServer = new GrpcServer(this.logger);
```
Comment (ImmanuelSegol) [inline lib/Walli.ts]: Should I start the gRPC server in the `Walli.start()` function?
Or should I initialize it in an express server like they do in xud?

Comment (michael1011) [reply lib/Walli.ts]: In `Walli.start()`. The express server is for using the gRPC with a HTTP client and we don't need this for now imo. @ImmanuelSegol

**Comment by michael1011** [comment]
Which operating system are you using? I also opened a PR that involves compiling protobufs for the proto file of LND and [my script looks the exact same as yours.](https://github.com/dopetard/walli-server/pull/41/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R32) Is there an error message or does the protobuf simply not show up? @ImmanuelSegol

**Comment by ImmanuelSegol** [comment]
@michael1011 Its working, please review 👍

File: lib/Walli.ts [hunk 2]
```diff
@@ -2,27 +2,31 @@ import Logger from './Logger';
...
+import GrpcServer from './grpc/GrpcServer';
+import Service from './service/Service';
 import { Arguments } from 'yargs';
 
 class Walli {
+  public service: Service;
   private config: ConfigType;
-  private logger: Logger;
-
+  private logger!: Logger;
```
Comment (michael1011) [inline lib/Walli.ts]: Why `logger!` instead of `logger`?

File: lib/Walli.ts [hunk 3]
```diff
@@ -54,7 +67,7 @@ class Walli {
     }
   }
 
-  private logCouldNotConnect = (service: string, error: any) => {
+  private logCouldNotConnect = (service: string, error?: any) => {
```
Comment (michael1011) [inline lib/Walli.ts]: `error?` should never be `undefined` or `null` and therefore it should stay `error`

File: lib/Walli.ts [hunk 4]
```diff
@@ -2,27 +2,31 @@ import Logger from './Logger';
...
+import Service from './service/Service';
 import { Arguments } from 'yargs';
 
 class Walli {
+  public service: Service;
   private config: ConfigType;
-  private logger: Logger;
-
+  private logger!: Logger;
+  private grpcServer?: GrpcServer;
```
Comment (michael1011) [inline lib/Walli.ts]: Why are `service` and `grpcServer` global variables?

File: lib/cli/Command.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import { Arguments } from 'yargs';
+import grpc from 'grpc';
+import { WalliClient } from '../proto/wallirpc_grpc_pb';
+
+export const loadWalliClient = (argv: Arguments): WalliClient => {
+  const credentials = grpc.credentials.createInsecure();
+  return new WalliClient(`${argv.rpc.host}:${argv.rpc.port}`, credentials);
+};
+
+interface GrpcResponse {
```
Comment (michael1011) [inline lib/cli/Command.ts]: Could you put the `interface` in between the imports and consts?

File: lib/cli/commands/getInfo.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { Arguments } from 'yargs';
+import { callback, loadWalliClient } from '../command';
+import { GetInfoRequest } from '../../proto/wallirpc_pb';
+export const command = 'getinfo';
+
+export const describe = 'get info about btcd';
```
Comment (michael1011) [inline lib/cli/commands/getInfo.ts]: Just returning information about BTCD is enough for testing but we should extend this command to show useful data about LND, BTCD and walli-server itself.

File: lib/grpc/GrpcServer.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    this.logger = logger;
+
+    const grpcService = new GrpcService(logger, service);
+    this.server.addService(WalliService, {
+      getInfo: grpcService.getInfo,
+    });
+  }
+
+  public listen = async (port: number, host: string) => {
+    return new Promise((resolve, reject) => {
```
Comment (michael1011) [inline lib/grpc/GrpcServer.ts]: Why are you returning a `Promise` here? When declaring a function as `async` you don't have to do this. `return` and `throw` will have the same effect as `resolve` and `reject`.

File: lib/grpc/errors.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+class GRPCError {
```
Comment (michael1011) [inline lib/grpc/errors.ts]: I would prefer a `type` over a class and what about renaming that `type` to `WalliError` and using it for all errors? I created some in `lib/consts/errors.ts`. Also why aren't you using an error code?

File: lib/service/Service.ts [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    const {version,
+          protocolversion,
+          blocks,
+          timeoffset,
+          connections,
+          proxy,
+          difficulty,
+          testnet,
+          relayfee } = info;
+    return {
```
Comment (michael1011) [inline lib/service/Service.ts]: The `const` in line 15 and this return statement seems really redundant to me.

File: proto/wallirpc.proto [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+message GetInfoRequest {}
+message GetInfoResponse {
+   int32 version = 1;
+   int32 protocolversion = 2;
+   int32 blocks = 3;
+   int32 timeoffset = 4;
+   int32 connections = 5;
+   string proxy = 6;
+   string difficulty = 7;
```
Comment (michael1011) [inline proto/wallirpc.proto]: Just out of curiosity: why isn't `difficulty` a `string` not an `int64`?

File: proto/wallirpc.proto [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+package wallirpc;
+
+service Walli {
+  rpc getInfo (GetInfoRequest) returns (GetInfoResponse);
+}
+
+message GetInfoRequest {}
+message GetInfoResponse {
+   int32 version = 1;
```
Comment (michael1011) [inline proto/wallirpc.proto]: Are the `[json_name = "version"]` in the proto file of XUD just used for the webproxy or is there a different reason for which you don't use them here?

File: lib/cli/commands/getInfo.ts (hunk 6 above)
Comment (ImmanuelSegol) [reply lib/cli/commands/getInfo.ts]: I think we should first look at issue #43.

File: proto/wallirpc.proto (hunk 11 above)
Comment (ImmanuelSegol) [reply proto/wallirpc.proto]: I didn't think we need it because we don't have a web proxy yet.

File: lib/grpc/errors.ts [hunk 12]
```diff
@@ -0,0 +1,13 @@
+import { Error } from '../consts/types';
+import errorCodePrefix from '../consts/errorCodePrefix';
+import { concatErrorCode } from '../Utils';
+
+/**
+ * Walli-Grpc specific errors
```
Comment (michael1011) [inline lib/grpc/errors.ts]: Please merge after removing this comment :+1:

**Pull request merged**